values/binary.go

```diff
@@ -22,9 +22,33 @@ func LookupBinaryFunction(sig BinaryFuncSignature) (BinaryFunction, error) {
 	if !ok {
 		return nil, fmt.Errorf("unsupported binary expression %v %v %v", sig.Left, sig.Operator, sig.Right)
 	}
-	return f, nil
+	return binaryFuncNullCheck(f), nil
 }
 
+// binaryFuncNullCheck will wrap any BinaryFunction and
+// check that both of the arguments are non-nil.
+//
+// If either value is null, then it will return null.
+// Otherwise, it will invoke the function to retrieve the result.
+func binaryFuncNullCheck(fn BinaryFunction) BinaryFunction {
+	return func(lv, rv Value) Value {
+		if lv.IsNull() || rv.IsNull() {
+			return Null
+		}
+		return fn(lv, rv)
+	}
+}
+
+// binaryFuncLookup contains a mapping of BinaryFuncSignature's to
+// the BinaryFunction that implements them.
+//
+// The values passed into these functions will be non-nil so a null
+// check is unnecessary inside of them.
+//
+// Even though nulls will never be passed to these functions,
+// the left or right type can be defined as nil. This is used to
+// mark that it is valid to use the operator between those two types,
+// but the function will never be invoked so it can be nil.
```

Comment:
Can we just remove these elements from the map since they'll never be called?

Reply:
It's to keep the binary lookup error message working. So if you try to do `int + null` it won't lookup and it will fail. Honestly, the ones for the comparison operators are going away as soon as we start working on implementing some form of typeclasses.